File: merge_sort_step.java
import java.io.*;
import java.util.*;

public class merge_sort_step {

    private static BufferedWriter stepWriter;

    // Data class holding integer + string pair
    static class Data {
        int key;
        String val;
        Data(int k, String v) { key = k; val = v; }
    }

    // Format array as [key/val, key/val, ...]
    private static String formatArray(Data[] arr) {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < arr.length; i++) {
            sb.append(arr[i].key).append("/").append(arr[i].val);
            if (i != arr.length - 1) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }

    // MergeSort entry
    public static void mergeSort(Data[] arr) throws IOException {
        mergeSortHelper(arr, 0, arr.length - 1);
    }

    // Recursive merge sort with step logging
    private static void mergeSortHelper(Data[] arr, int left, int right) throws IOException {
        if (left >= right) return;
        int mid = left + (right - left) / 2;
        mergeSortHelper(arr, left, mid);
        mergeSortHelper(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }

    // Merge two subarrays and log the steps
    private static void merge(Data[] arr, int left, int mid, int right) throws IOException {
        Data[] leftArr = Arrays.copyOfRange(arr, left, mid + 1);
        Data[] rightArr = Arrays.copyOfRange(arr, mid + 1, right + 1);

        // Log left and right arrays before merging
        stepWriter.write(formatArray(leftArr) + "\n");
        stepWriter.write(formatArray(rightArr) + "\n");

        int i = 0, j = 0, k = left;
        while (i < leftArr.length && j < rightArr.length) {
            if (leftArr[i].key <= rightArr[j].key) {
                arr[k++] = leftArr[i++];
            } else {
                arr[k++] = rightArr[j++];
            }
        }
        while (i < leftArr.length) arr[k++] = leftArr[i++];
        while (j < rightArr.length) arr[k++] = rightArr[j++];

        // Log the merged array
        stepWriter.write("Merged: " + formatArray(Arrays.copyOfRange(arr, left, right + 1)) + "\n\n");
    }

    // Read CSV data and return it as Data[]
    private static List<Data> readCSV(String filename, int startRow, int endRow) throws IOException {
        List<Data> list = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            int row = 1;
            while ((line = br.readLine()) != null) {
                if (row >= startRow && row <= endRow) {
                    String[] parts = line.split(",", 2);
                    if (parts.length == 2) {
                        try {
                            int key = Integer.parseInt(parts[0].trim());
                            String val = parts[1].trim();
                            list.add(new Data(key, val));
                        } catch (NumberFormatException e) {
                            System.err.println("Skipping malformed integer at line " + row);
                        }
                    } else {
                        System.err.println("Skipping malformed line at row " + row);
                    }
                }
                if (row > endRow) break;
                row++;
            }
        }
        return list;
    }

    public static void main(String[] args) {
        if (args.length < 3) {
            System.out.println("Usage: java merge_sort_step <csv_file> <start_row> <end_row>");
            return;
        }

        String csvFile = args[0];
        int startRow, endRow;

        try {
            startRow = Integer.parseInt(args[1]);
            endRow = Integer.parseInt(args[2]);
        } catch (NumberFormatException e) {
            System.err.println("Start and end rows must be integers.");
            return;
        }

        try {
            List<Data> dataList = readCSV(csvFile, startRow, endRow);
            if (dataList.isEmpty()) {
                System.err.println("No data found in the specified row range.");
                return;
            }
            Data[] arr = dataList.toArray(new Data[0]);

            String outFile = String.format("merge_sort_step_%d_%d.txt", startRow, endRow);
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(outFile))) {
                stepWriter = writer;
                stepWriter.write("# Merge sort steps and sorted data\n");
                stepWriter.write("Original Array:," + formatArray(arr) + "\n");
                mergeSort(arr);
            }

            System.out.println("Merge sort steps saved to " + outFile);

        } catch (IOException e) {
            System.err.println("Error:");
            e.printStackTrace();
        }
    }
}

File: merge_sort_step.py
import sys

class Data:
    def __init__(self, key, val):
        self.key = key
        self.val = val
    
    def __str__(self):
        return f"{self.key}/{self.val}"

def format_array(arr):
    return "[" + ", ".join(str(x) for x in arr) + "]"

def merge(arr, left, mid, right, step_file):
    leftArr = arr[left:mid+1]
    rightArr = arr[mid+1:right+1]

    i = j = 0
    k = left

    while i < len(leftArr) and j < len(rightArr):
        if leftArr[i].key <= rightArr[j].key:
            arr[k] = leftArr[i]
            i += 1
        else:
            arr[k] = rightArr[j]
            j += 1
        k += 1

    while i < len(leftArr):
        arr[k] = leftArr[i]
        i += 1
        k += 1

    while j < len(rightArr):
        arr[k] = rightArr[j]
        j += 1
        k += 1

    # Log the entire array state after this merge
    step_file.write(format_array(arr) + "\n")

def merge_sort_helper(arr, left, right, step_file):
    if left >= right:
        return
    mid = (left + right) // 2
    merge_sort_helper(arr, left, mid, step_file)
    merge_sort_helper(arr, mid + 1, right, step_file)
    merge(arr, left, mid, right, step_file)

def merge_sort(arr, step_file):
    if len(arr) < 2:
        return
    merge_sort_helper(arr, 0, len(arr) - 1, step_file)

def read_csv(filename, start_row, end_row):
    data = []
    with open(filename, 'r') as f:
        for idx, line in enumerate(f, 1):
            if idx < start_row:
                continue
            if idx > end_row:
                break
            parts = line.strip().split(',', 1)  # Split into max 2 parts
            if len(parts) == 2:
                try:
                    key = int(parts[0])
                    val = parts[1].strip()
                    data.append(Data(key, val))
                except ValueError:
                    print(f"Skipping malformed integer at line {idx}")
    return data

def main():
    if len(sys.argv) < 4:
        print("Usage: python merge_sort_step.py <csv_file> <start_row> <end_row>")
        sys.exit(1)

    csv_file = sys.argv[1]
    start_row = int(sys.argv[2])
    end_row = int(sys.argv[3])

    arr = read_csv(csv_file, start_row, end_row)

    if not arr:
        print("No data read from the CSV file within the specified rows.")
        sys.exit(1)

    out_file = f"merge_sort_step_{start_row}_{end_row}.txt"
    with open(out_file, 'w') as step_file:
        step_file.write("# Merge sort steps and sorted data\n")
        step_file.write("Original Array:," + format_array(arr) + "\n")
        merge_sort(arr, step_file)

    print(f"Merge steps saved to {out_file}")

if __name__ == "__main__":
    main()
